Scripts/Runtime/Jobs/Collections/AccessController.cs

```diff
@@ -0,0 +1,141 @@
+using Unity.Jobs;
+
+namespace Anvil.Unity.DOTS.Jobs
+{
+    /// <summary>
+    /// A utility class that handles managing access for reading and writing so that jobs can be scheduled easily.
+    /// </summary>
+    /// <remarks>
+    /// This can be a little bit complicated to wrap your head around so here is an example
+    ///
+    /// ----- EXTERNAL WRITE PHASE -----
+    /// - Multiple different systems want to write to the collection.
+    /// - They schedule their writing jobs using the <see cref="WriteAccessDependency"/> handle.
+    /// - This means that all those jobs can start at the same time.
+    /// - All of those jobs use <see cref="AddJobHandleForParallelWriting"/> to let the <see cref="AccessController"/> know that there is writing going on. We cannot read until this is done.
+    ///
+    /// ----- INTERNAL WRITE PHASE -----
+    /// - A managing system now needs to do some work where it reads from and writes to the collection.
+    /// - It schedules it's job to do that using the <see cref="ReadWriteAccessDependency"/> handle.
+    /// - This means that it can do it's work once all the previous external writers have completed.
+    /// - This job then uses <see cref="AddJobHandleForReadWriting"/> to the let the <see cref="AccessController"/> know that there is reading and writing going on that cannot be interrupted by external parallel writes.
+    ///
+    /// ----- EXTERNAL READ PHASE -----
+    /// - Multiple different systems want to read from the collection.
+    /// - They schedule their reading jobs using the <see cref="ReadAccessDependency"/> handle.
+    /// - This means that all those jobs can start at the same time.
+    /// - All of those jobs use <see cref="AddJobHandleForParallelReading"/> to let the <see cref="AccessController"/> know that there is reading going on. We cannot write until this is done.
+    ///
+    /// ----- CLEAN UP PHASE -----
+    /// - The collection used above needs to be disposed but we need to ensure all reading and writing are complete.
+    /// - The collection disposes using the <see cref="ParallelReadAccessDependency"/> handle.
```

Comment:
This is a good point too, I've reworked things a bit so that Disposal is its own `AccessType` called `ForDisposal`

You get `ExclusiveWrite` access at the end of the chain and no one is allowed to `Acquire` again afterwards. 

In the case of management systems where you are creating/destroying collections every frame, you can call `Reset` to keep the same instance but reset the access controller to reflect the new collection. 

Here's an excerpt from the `TimerSubSystem` that demonstrates:

```
            //The previous frame's completed timers can be disposed once all the reads are complete
            m_CompletedTimers.Dispose(m_CollectionAccessController.Acquire(AccessType.ForDisposal));
            
            //We don't need to release because we were disposing but we do want to reset the controller now since 
            //we have new work on a new collection.
            m_CollectionAccessController.Reset();

            //We then create a new completed buffer to write any completed timers for this frame
            m_CompletedTimers = new UnsafeTypedStream<Timer>(Allocator.TempJob);
```